src/main/java/org/broadinstitute/hellbender/tools/exome/CopyNumberTriStateSegmentCaller.java

```diff
@@ -0,0 +1,362 @@
+        final TargetCollection<Target> targets = targetArguments.readTargetCollection(false);
+        final ReadCountCollection inputCounts = readAndSortCountsByTargetCoordinates(targets);
+        applyZScoreTransformation(inputCounts);
+        checkForMissingTargetsInInputCounts(targets, inputCounts);
+        openOutput(outputFile, model, targets, inputCounts);
+        final Object result = makeCalls(model, targets, inputCounts);
+        closeOutput(outputFile);
+        return result;
+    }
+
+    /**
+     * Opens the output where to dump the calls.
+     *
+     * @throws UserException.CouldNotCreateOutputFile if something went wrong creating/opening or writing in the file.
+     * @throws UserException.BadInput if there inputs (coverage or targets) are somehow invalid.
+     * @throws IllegalArgumentException if any of the parameters is {@code null}.
+     */
+    protected abstract void openOutput(final File outputFile, final CopyNumberTriStateHiddenMarkovModel model,
+                                       final TargetCollection<Target> targets, final ReadCountCollection inputCounts);
+
+    /**
+     * Closes the output.
+     *
+     * @throws UserException.CouldNotCreateOutputFile if something went wrong.
+     */
+    protected abstract void closeOutput(final File outputFile);
+
+    /**
+     * Does the actual CN segment calling, outputting results to the output file.
+     *
+     * @param model the CN segment model.
+     * @param targets the target under analysis.
+     * @param inputCounts the input coverage values already processed.
+     * @return IllegalArgumentException if any of the parameters is {@code null} or otherwise invalid.
+     * @return IllegalStateException if the tool is no ready for calling (e.g. the output file was not properly opened).
+     * @return UserException if anything when wrong due to something under the users' control.
+     * @return GATKException if anything when wrong due to something outside the users' control.
+     */
+    protected abstract Object makeCalls(final CopyNumberTriStateHiddenMarkovModel model,
+                                        final TargetCollection<Target> targets,
+                                        final ReadCountCollection inputCounts);
+
+    /**
+     * Transform read counts to z-scores.
+     * @param inputCounts the input read-counts, modified in-situ.
+     */
+    private final void applyZScoreTransformation(final ReadCountCollection inputCounts) {
+        final RealMatrix counts = inputCounts.counts();
+        switch (zscoreDimension) {
+            case SAMPLE:
+                standardizeBySample(counts);
+                break;
+            case TARGET:
+                standardizeByTarget(counts);
+                break;
+        }
+    }
+
+    private final void standardizeByTarget(final RealMatrix counts) {
+        final double[] rowMeans = IntStream.range(0, counts.getRowDimension())
```

Comment:
Need to use the GATKProtectedMathUtils.columnMeans, StdDev, rowMeans, StDev methods. This changed were lost when rebasing.